parcels/kernelsoa.py

```diff
@@ -0,0 +1,370 @@
+
+
+__all__ = ['KernelSOA']
+
+
+class KernelSOA(BaseKernel):
+    """Kernel object that encapsulates auto-generated code.
+
+    :arg fieldset: FieldSet object providing the field information
+    :arg ptype: PType object for the kernel particle
+    :param delete_cfiles: Boolean whether to delete the C-files after compilation in JIT mode (default is True)
+
+    Note: A Kernel is either created from a compiled <function ...> object
+    or the necessary information (funcname, funccode, funcvars) is provided.
+    The py_ast argument may be derived from the code string, but for
+    concatenation, the merged AST plus the new header definition is required.
+    """
+
+    def __init__(self, fieldset, ptype, pyfunc=None, funcname=None,
+                 funccode=None, py_ast=None, funcvars=None, c_include="", delete_cfiles=True):
+        super(KernelSOA, self).__init__(fieldset=fieldset, ptype=ptype, pyfunc=pyfunc, funcname=funcname, funccode=funccode, py_ast=py_ast, funcvars=funcvars, c_include=c_include, delete_cfiles=delete_cfiles)
+
+        # Derive meta information from pyfunc, if not given
+        self.check_fieldsets_in_kernels(pyfunc)
+
+        if funcvars is not None:
+            self.funcvars = funcvars
+        elif hasattr(pyfunc, '__code__'):
+            self.funcvars = list(pyfunc.__code__.co_varnames)
+        else:
+            self.funcvars = None
+        self.funccode = funccode or inspect.getsource(pyfunc.__code__)
+        # Parse AST if it is not provided explicitly
+        self.py_ast = py_ast or parse(BaseKernel.fix_indentation(self.funccode)).body[0]
+        if pyfunc is None:
+            # Extract user context by inspecting the call stack
+            stack = inspect.stack()
+            try:
+                user_ctx = stack[-1][0].f_globals
+                user_ctx['math'] = globals()['math']
+                user_ctx['ParcelsRandom'] = globals()['ParcelsRandom']
+                user_ctx['random'] = globals()['random']
+                user_ctx['StateCode'] = globals()['StateCode']
+                user_ctx['OperationCode'] = globals()['OperationCode']
+                user_ctx['ErrorCode'] = globals()['ErrorCode']
+            except:
+                logger.warning("Could not access user context when merging kernels")
+                user_ctx = globals()
+            finally:
+                del stack  # Remove cyclic references
+            # Compile and generate Python function from AST
+            py_mod = parse("")
+            py_mod.body = [self.py_ast]
+            exec(compile(py_mod, "<ast>", "exec"), user_ctx)
+            self._pyfunc = user_ctx[self.funcname]
+        else:
+            self._pyfunc = pyfunc
+
+        numkernelargs = self.check_kernel_signature_on_version()
+
+        assert numkernelargs == 3, \
+            'Since Parcels v2.0, kernels do only take 3 arguments: particle, fieldset, time !! AND !! Argument order in field interpolation is time, depth, lat, lon.'
```

Comment:
I imagine that this code (but also a lot of the rest in this `__init__` is very general and would apply to _all_ kernels? Could we move it to the BaseKernel? Or is that technically difficult?

Reply:
nope, we can't because of the instruction flow. Meaning: this kernel-arg number require's the pyfunc to be initialised. That would though require some kernel-specific interpretations.

We can only put things in the `BaseKernel`'s `__ini__` function that is completely (i.e. provably - not just _potentially_) function-independent. That is also why the `pyfunc` is interpreted in the specific- and not the base class.

Specifically: we have those `use_ctx` commands. If we have a structure (e.g. nodes) that require a multi-object, multi-stage compilation, then there are some context-commands that change (e.g. around the `random` context). That makes this `pyfunc` initialisation kernel-/structure-dependent, and hence everything that relies on it needs to follow in the derived class.

Reply:
As you see, in order to get around the problem, the actual check to obtain the `numkernelargs` has been moved to a member function inside the base class - which is the way to go for those cases.